codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.redis;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.CodecException;
+import io.netty.handler.codec.MessageToMessageEncoder;
+import io.netty.util.CharsetUtil;
+import io.netty.util.ReferenceCountUtil;
+
+import java.nio.CharBuffer;
+import java.util.List;
+
+/**
+ * Encodes {@link RedisMessage} into bytes following
+ * <a href="http://redis.io/topics/protocol">RESP (REdis Serialization Protocol)</a>.
+ */
+public class RedisEncoder extends MessageToMessageEncoder<RedisMessage> {
+
+    private static final int TYPE_LENGTH = 1;
+    private static final byte[] CRLF = { '\r', '\n' };
+    private static final byte[] NULL_BULK_STRING = { '-', '1' };
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, RedisMessage msg, List<Object> out) throws Exception {
+        ByteBuf buf = null;
+        try {
+            buf = ctx.alloc().buffer(calculateBufferLength(msg));
```

Comment:
The suggestion in https://github.com/netty/netty/pull/5001#discussion_r57942794 (don't aggregate at the lowest level) would impact how encoding is done ... however I'm wondering why we have to allocate a buffer and copy into it for every encoded message. Seems like we should be able to add the parts directly to `out`, and thus not have to worry about `calculateBufferLength`...


Here is an example of what the `writeBulkStringMessage` may look like...
```java
private static final ByteBuf NULL_BULK_STRING = Unpooled.unmodifiableBuffer(Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(new byte[] {'-', '1'})));
private static final ByteBuf EOL = Unpooled.unmodifiableBuffer(Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(new byte[] {'\r', '\n'})));

private static void writeBulkStringMessage(List<Object> out, ByteBufAllocator alloc, BulkStringRedisMessage msg) {
  if (msg instanceof BulkStringRedisMessage) {
    // This buffer should be cached/unreleasble/unmodifiable like EOL and a slice should be taken
    out.add(alloc.ioBuffer(1).writeLong(msg.type().value()));
    if (msg.isNull()) {
      out.add(NULL_BULK_STRING.slice());
    } else {
      // We could use the Integer ByteBuf cache below
      out.add(alloc.ioBuffer(65).writeBytes(msg.expectedBytes()));
      out.add(EOL.slice());
    }
  }

  if (msg instanceof LastBulkStringRedisContent) {
    out.add(msg.content());
  }
  if (msg instanceof LastBulkStringRedisMessage) {
    out.add(EOL.slice());
  }
}
```